### PR TITLE
Allow references to intermediate IDs after multiple name transformations

### LINKED ISSUE #3228: [Question] Is a var required when names are ambiguous?
I'm using the same component in two resources but need to refer to hashed Secret and ConfigMap name from both resources in my project. Do I need to use a var to get the correct Secret/ConfigMap name + hash if the name reference is ambiguous or am I missing something?

Example:

I've tried to boil my issue down to a bare minimum. xql-zero & xql-one inherit from xql, app references xql-zero and xql-one

A rough Helm Chart equivalent would be:

```
apiVersion: "v2"
description: Helm Chart for "app"
name: app
version: 0.0.1
appVersion: "0.0.1-initial"

dependencies:
- name: xql
  version: 0.0.1
  alias: xql-zero
- name: xql
  version: 0.0.1
  alias: xql-one
```

Ex:

Note: `XQL_ONE_PASSWORD` and `XQL_ZERO_PASSWORD`, which are not referenced using vars, have a keyref name of `xql-secret` while `XQL_ZERO_ARG` and `XQL_ONE_ARG`, which use vars, have hashed names: `project-xql-zero-xql-secret-6khmtc56hm`

`XQL_ONE_ARG` (with var):
```
- name: XQL_ONE_ARG
  valueFrom:
    secretKeyRef:
      name: $(PROJECT_XQL_ONE_SECRET_NAME) # Defined below in "INPUT"
      key: arg
```

`XQL_ONE_PASSWORD` (without var):
```
- name: XQL_ONE_PASSWORD
  valueFrom:
    secretKeyRef:
      name: xql-secret
      key: password
```

OUTPUT:
```
Error: obj '{"apiVersion": "apps/v1", "kind": "Deployment", "metadata": {"name": "project-app"},
  "spec": {"template": {"spec": {"containers": [{"env": [{"name": "XQL_ZERO_ARG",
                "valueFrom": {"secretKeyRef": {"key": "arg", "name": "$(PROJECT_XQL_ZERO_SECRET_NAME)"}}},
              {"name": "XQL_ZERO_PASSWORD", "valueFrom": {"secretKeyRef": {"key": "password",
                    "name": "xql-secret"}}}, {"name": "XQL_ONE_ARG", "valueFrom": {
                  "secretKeyRef": {"key": "arg", "name": "$(PROJECT_XQL_ONE_SECRET_NAME)"}}},
              {"name": "XQL_ONE_PASSWORD", "valueFrom": {"secretKeyRef": {"key": "password",
                    "name": "xql-secret"}}}], "image": "example.com/app:latest", "imagePullPolicy": "Always",
            "name": "app"}]}}}}
' at path 'spec/template/spec/containers/env/valueFrom/secretKeyRef/name': visit traversal on path: [env valueFrom secretKeyRef name]: visit traversal on path: [valueFrom secretKeyRef name]: multiple matches for ~G_v1_Secret|~X|xql-secret:
  [~G_v1_Secret|~X|project-xql-zero-xql-secret-6khmtc56hm
 ~G_v1_Secret|~X|project-xql-one-xql-secret-79mhmf5dgt
]

```

INPUT:

[wat.zip](https://github.com/kubernetes-sigs/kustomize/files/5542879/wat.zip)

```
==> ./kustomization.yaml <==
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

namePrefix: project-

resources:
- app
==> ./app/kustomization.yaml <==
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

resources:
- resources/deployment.yaml
- resources/xql
==> ./app/resources/xql/kustomization.yaml <==
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

resources:
- xql-zero
- xql-one

configurations:
- ./kustomizeconfig.yaml

==> ./app/resources/xql/kustomizeconfig.yaml <==
varReference:
- path: spec/template/spec/containers/env/valueFrom/secretKeyRef/name
==> ./app/resources/xql/xql-one/kustomization.yaml <==
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

namePrefix: xql-one-

resources:
- ../../../../../bases/xql

secretGenerator:
- name: xql-secret
  behavior: merge
  envs:
  - config/xql-one-secret.env

vars:
- name: PROJECT_XQL_ONE_SECRET_NAME
  objref:
    kind: Secret
    name: xql-secret
    apiVersion: v1
  fieldref:
    fieldpath: metadata.name

==> ./app/resources/xql/xql-one/config/xql-one-secret.env <==
arg=1
==> ./app/resources/xql/xql-zero/kustomization.yaml <==
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

namePrefix: xql-zero-

resources:
- ../../../../../bases/xql

secretGenerator:
- name: xql-secret
  behavior: merge
  envs:
  - config/xql-zero-secret.env

vars:
- name: PROJECT_XQL_ZERO_SECRET_NAME
  objref:
    kind: Secret
    name: xql-secret
    apiVersion: v1
  fieldref:
    fieldpath: metadata.name

==> ./app/resources/xql/xql-zero/config/xql-zero-secret.env <==
arg=0
==> ./app/resources/deployment.yaml <==
apiVersion: apps/v1
kind: Deployment
metadata:
  name: app
spec:
  template:
    spec:
      containers:
      - name: app
        image: example.com/app:latest
        imagePullPolicy: Always
        env:
        - name: XQL_ZERO_ARG
          valueFrom:
            secretKeyRef:
              name: $(PROJECT_XQL_ZERO_SECRET_NAME)
              key: arg
        - name: XQL_ZERO_PASSWORD
          valueFrom:
            secretKeyRef:
              name: xql-secret
              key: password
        - name: XQL_ONE_ARG
          valueFrom:
            secretKeyRef:
              name: $(PROJECT_XQL_ONE_SECRET_NAME)
              key: arg
        - name: XQL_ONE_PASSWORD
          valueFrom:
            secretKeyRef:
              name: xql-secret
              key: password
```

Referring to the two `xql-secert`-s exclusively using the vars `PROJECT_XQL_ZERO_SECRET_NAME` and `PROJECT_XQL_ONE_SECRET_NAME` produces the expected output, but, again, is hacky.

INPUT WITH VARS:

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: app
spec:
  template:
    spec:
      containers:
      - name: app
        image: example.com/app:latest
        imagePullPolicy: Always
        env:
        - name: XQL_ZERO_ARG
          valueFrom:
            secretKeyRef:
              name: $(PROJECT_XQL_ZERO_SECRET_NAME)
              key: arg
        - name: XQL_ZERO_PASSWORD
          valueFrom:
            secretKeyRef:
              name: $(PROJECT_XQL_ZERO_SECRET_NAME)
              key: password
        - name: XQL_ONE_ARG
          valueFrom:
            secretKeyRef:
              name: $(PROJECT_XQL_ONE_SECRET_NAME)
              key: arg
        - name: XQL_ONE_PASSWORD
          valueFrom:
            secretKeyRef:
              name: $(PROJECT_XQL_ONE_SECRET_NAME)
              key: password
```

OUTPUT WITH VARS:

```
apiVersion: v1
data:
  arg: MQ==
  password: U1VQRVJfU0VDUkVUX1BBU1NXT1JE
kind: Secret
metadata:
  annotations: {}
  labels: {}
  name: project-xql-one-xql-secret-79mhmf5dgt
type: Opaque
---
apiVersion: v1
data:
  arg: MA==
  password: U1VQRVJfU0VDUkVUX1BBU1NXT1JE
kind: Secret
metadata:
  annotations: {}
  labels: {}
  name: project-xql-zero-xql-secret-6khmtc56hm
type: Opaque
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: project-app
spec:
  template:
    spec:
      containers:
      - env:
        - name: XQL_ZERO_ARG
          valueFrom:
            secretKeyRef:
              key: arg
              name: project-xql-zero-xql-secret-6khmtc56hm
        - name: XQL_ZERO_PASSWORD
          valueFrom:
            secretKeyRef:
              key: password
              name: project-xql-zero-xql-secret-6khmtc56hm
        - name: XQL_ONE_ARG
          valueFrom:
            secretKeyRef:
              key: arg
              name: project-xql-one-xql-secret-79mhmf5dgt
        - name: XQL_ONE_PASSWORD
          valueFrom:
            secretKeyRef:
              key: password
              name: project-xql-one-xql-secret-79mhmf5dgt
        image: example.com/app:latest
        imagePullPolicy: Always
        name: app
```

### LINKED ISSUE #3455: Allow references to intermediate names after multiple name transformations
For example, if a user has this stack: 

overlay: add a prefix “gcp-”
overlay: add a prefix ‘emea-’
overlay: add a prefix ‘prod-’
base: deployment named ‘foo’

The gcp layer can use the name “foo” or “emea-prod-foo”, but not just the name “prod-foo”. The gcp layer should be able to refer to the intermediate name “prod-foo” as well.

### PR DESCRIPTION
**Pull request by natasha41575**
Fix #3228 
Fix #3455

### CODE REVIEW

**Comment by natasha41575** [comment]
@Shell32-Natsu @monopole please take a look

**Comment by monopole** [comment]
Existing methods involved in name and namespace changes are
```
  Resource
    AddNamePrefix / Suffix
    CurId / OrgId
    GetNamePrefixes / Suffixes
    GetOriginalName / Ns
    GetOutermostNamePrefix / Suffix
    OutermostPrefixSuffixEquals
    PrefixesSuffixesEquals
    SetName / Namespace
    SetOriginalName / Ns

  ResMap
    GetByCurrentId
    GetById
    GetByOriginalId
    GetIndexOfCurrentId
    GetMatchingResourcesByCurrentId
    GetMatchingResourcesByOriginalId
    GroupedByCurrentNamespace
    GroupedByOriginalNamespace

  annotations
    buildAnnotationOriginalName
    buildAnnotationPrefixes
    buildAnnotationSuffixes
    buildAnnotationOriginalNamespace
```

This PR leaves that alone and adds the additional methods
```
  Resource
    addIntermediateName / Ns
    getIntermediateNames / Ns
    PreviouslyNamed
    PrevIds
    StorePreviousId

  ResMap
    GetByIntermediateId
    GetMatchingResourcesByIntermediateId

  annotations
    buildAnnotationIntermediateNames
    buildAnnotationIntermediateNs
```

These are interesting additions, suggesting a better GVKNNs api
that boils down to

  GetCurrentId -> ResId
  GetPreviousIds -> ResId[]

This PR does what it purports to do, but is large and increases code complexity - new annotations, new API, nothing removed or simplified.

Too much in existing the Resource API is exposing prefix/suffix stuff.  Some of it doesn't appear to be used anymore.

If we're going down a road that allows arbitrary GVKN changes,
it's gonna get worse.

I'd like to identify some ways to simplify too.


Some specific things:


 - New public methods need a regression test.
   Cover no name changes, one name change and two name changes.

 - The comment for PrevId should be clear on what it returns
   if the resource hasn't had a name change.
   nil, empty list, or the current Id?  If its true that PrevIds
   never returns the CurrentId, then the comment should just
   say that.

 -
Misc

 - please rename
     addAdditiveAnnotation -> appendCsvAnnotation
     getAdditiveAnnotation -> getCsvAnnotations

 - use existing const resid.DefaultNamespace rather than introducing a new const.

**Comment by natasha41575** [comment]
@monopole Thanks for your refactoring, it made this a lot simpler and easier to implement. It is ready for review when you have the chance to take a look

**Review by Shell32-Natsu — COMMENTED** [review]
Overall LGTM. Thanks!

Only one question: what if different resources in different layers have the same name?

**Comment by natasha41575** [comment]
> Overall LGTM. Thanks!
> 
> Only one question: what if different resources in different layers have the same name?

Added a couple tests for what I think is the scenario you are describing

File: api/krusty/intermediateresourceid_test.go [hunk 1]
```diff
@@ -67,6 +68,147 @@ spec:
...
+metadata:
+  name: foo
+spec:
+  template:
+    spec:
+      containers:
+      - image: whatever
+`)
+	err := th.RunWithErr("/app/gcp", th.MakeDefaultOptions())
+	assert.Error(t, err)
```
Comment (Shell32-Natsu) [inline api/krusty/intermediateresourceid_test.go:209]: What does the error message look like? I am thinking that if a user has a lot of layers and duplicate names the error message will be very hard to read.

Comment (natasha41575) [reply api/krusty/intermediateresourceid_test.go:209]: Error looks like:
`multiple matches for Id apps_v1_Deployment|~X|foo; failed to find unique target for patch apps_v1_Deployment|foo`

Comment (Shell32-Natsu) [reply api/krusty/intermediateresourceid_test.go:209]: OK, let's go with it.

Comment (monopole) [reply api/krusty/intermediateresourceid_test.go:209]: sgtm too

**Review by monopole — APPROVED** [review]
/woot
//lgtm

**Comment by monopole** [comment]
starting release.

**Pull request merged**